openedx/core/djangoapps/profile_images/images.py

```diff
@@ -0,0 +1,130 @@
+"""
+Image file manipulation functions related to profile images.
+"""
+from cStringIO import StringIO
+
+from django.conf import settings
+from django.core.files.base import ContentFile
+from PIL import Image
+
+from ..user_api.accounts.api import get_profile_image_storage
+
+
+class DevMsg(object):
+    """
+    Holder for pseudo-constants.
+    """
+    FILE_TOO_LARGE = 'Maximum file size exceeded.'
+    FILE_TOO_SMALL = 'Minimum file size not met.'
+    FILE_BAD_TYPE = 'Unsupported file type.'
+    FILE_BAD_EXT = 'File extension does not match data.'
+    FILE_BAD_MIMETYPE = 'Content-Type header does not match data.'
+
+
+class ImageValidationError(Exception):
+    """
+    Exception to use when the system rejects a user-supplied source image.
+    """
+    pass
+
+
+def validate_uploaded_image(uploaded_file):
+    """
+    Raises ImageFileRejected if the server should refuse to use this
+    uploaded file as the source image for a user's profile image.  Otherwise,
+    returns nothing.
+    """
+    # validation code by @pmitros,
+    # adapted from https://github.com/pmitros/ProfileXBlock
+    # TODO: investigate if PIL has builtin methods for this
+
+    image_types = {
+        'jpeg': {
+            'extension': [".jpeg", ".jpg"],
+            'mimetypes': ['image/jpeg', 'image/pjpeg'],
+            'magic': ["ffd8"]
+        },
+        'png': {
+            'extension': [".png"],
+            'mimetypes': ['image/png'],
+            'magic': ["89504e470d0a1a0a"]
+        },
+        'gif': {
+            'extension': [".gif"],
+            'mimetypes': ['image/gif'],
+            'magic': ["474946383961", "474946383761"]
+        }
+    }
+
+    if uploaded_file.size > settings.PROFILE_IMAGE_MAX_BYTES:
+        raise ImageValidationError(DevMsg.FILE_TOO_LARGE)
+    elif uploaded_file.size < settings.PROFILE_IMAGE_MIN_BYTES:
+        raise ImageValidationError(DevMsg.FILE_TOO_SMALL)
+
+    # check the file extension looks acceptable
+    filename = str(uploaded_file.name).lower()
+    filetype = [ft for ft in image_types if any(filename.endswith(ext) for ext in image_types[ft]['extension'])]
+    if not filetype:
+        raise ImageValidationError(DevMsg.FILE_BAD_TYPE)
+    filetype = filetype[0]
+
+    # check mimetype matches expected file type
+    if uploaded_file.content_type not in image_types[filetype]['mimetypes']:
+        raise ImageValidationError(DevMsg.FILE_BAD_MIMETYPE)
+
+    # check image file headers match expected file type
+    headers = image_types[filetype]['magic']
+    if uploaded_file.read(len(headers[0]) / 2).encode('hex') not in headers:
```

Comment:
`.encode('hex')` is what fixes things - a single byte read from a file converted to a hex will result in a string of length 2.

the first magic number is chosen with the assumption that when multiple magic numbers may exist for a given file format, they will be the same length.

altogether, the point is to read exactly the number of bytes needed from the top of the file in order to make a string comparison against any of the possible valid magic numbers for the expected type of image file.

I'm putting an inline comment in the code with a link to http://en.wikipedia.org/wiki/Magic_number_%28programming%29  which I used myself as a quick reference to ensure I understood what this code was doing...